### PR TITLE
Added exec/eval

### PR DESCRIPTION
**Pull request by klemens-morgenstern**
This adds the exec/eval functions, blatantly copied from boost.python. I hope that fits into this library.

### CODE REVIEW

**Comment by wjakob** [comment]
Generally, this would be a useful feature to heave.

A few notes/thoughts:

1. Can we just copy stuff from Boost? It uses a different license.

2. This code does not follow the formatting conventions of the library (no tabs, indent with 4 spaces, brace placement, etc. etc.)

3. Documentation is too limited. Also, embedding the interpreter is not currently a supported feature, so it's weird that ``Py_Initialize() ;`` is used.

4. I would prefer not to have "exec" be a default include in pybind11.h

5. Testing is still insufficient, some methods are not tested at all (e.g. exec file). It would also be good to cover tricky cases like a file that is missing, or an exception being throw in the tests.

Thanks,
Wenzel

**Comment by klemens-morgenstern** [comment]
1. I think so, it grants the same permits, though the license code is not provided liberally - but the content is the same. I think this is sufficient.

2. Alright I'll fix that.

3. Well, yeah, but that's part of the feature. I thought about wrapping it into a wrapper, i.e. something like `py::session`. But it's too simple for this, I think. 

4. Sure, no problem.

5. Uhm, exec_file is tested, But sure, failure should be tested, I'll do that.

**Comment by klemens-morgenstern** [comment]
Ok 2, 4 and 5 are done. I replaced the tabs with spaces and changed the scope formatting

As for (1) Should I ask in the boost mailing list of we can copy that?

**Comment by dean0x7d** [comment]
In my experience with boost.python, the `exec_file` implementation on Python >= 3 leaks the file handle: `FILE* fs` is never closed.

You may want to use [`PyRun_FileEx`](https://docs.python.org/3.5/c-api/veryhigh.html#c.PyRun_FileEx) with `closeit = 1` on Python 3 and `closeit = 0` on Python 2. (The Python 2 version already closes the file thanks to `object file(pyfile, false);`)

BTW I don't know much about the legal obligations here, but it seems like the original boost license and copyright holders should be mentioned in the copied code.

**Comment by klemens-morgenstern** [comment]
@dean0x7d Ok, did it as you suggested. I also changed the way the fopen error is handled, so you always get an std::invalid_argument, when the file does not exist. I also test with python 2.7 locally, so that's covered.

File: include/pybind11/exec.h [hunk 1]
```diff
@@ -75,36 +77,46 @@ inline object exec_statement (str string, object global = object(), object local
 }
 
 inline object exec_file(str filename, object global = object(), object local = object()) {
-  // Set suitable default values for global and local dicts.
-  if (!global)
-  {
+    // Set suitable default values for global and local dicts.
+    if (!global)
+    {
```
Comment (wjakob) [inline include/pybind11/exec.h]: In pybind11, braces are on the same line as the previous statement.

File: include/pybind11/exec.h [hunk 2]
```diff
@@ -75,36 +77,46 @@ inline object exec_statement (str string, object global = object(), object local
...
+    if (!local) local = global;
+    // should be 'char const *' but older python versions don't use 'const' yet.
 
     auto f = static_cast<std::string>(filename);
 
-  // Let python open the file to avoid potential binary incompatibilities.
+    // Let python open the file to avoid potential binary incompatibilities.
 #if PY_VERSION_HEX >= 0x03040000
-  FILE *fs = _Py_fopen(f.c_str(), "r");
+    constexpr static int close_it = 1;
```
Comment (wjakob) [inline include/pybind11/exec.h]: constexpr static int -> const int.

File: include/pybind11/exec.h [hunk 3]
```diff
@@ -3,6 +3,8 @@
 
     Copyright (c) 2016 Klemens D. Morgenstern <klemens.morgenstern@gmx.net>
 
+    This code is based on the boost.python implementation.
```
Comment (wjakob) [inline include/pybind11/exec.h]: Need to specify full boost license here. I just checked, and the license specifically requires this.

File: example/example18.cpp [hunk 4]
```diff
@@ -0,0 +1,129 @@
+/*
```
Comment (wjakob) [inline example/example18.cpp]: Please make example18 a normal part of the extension library, just like example1-17.

File: CMakeLists.txt [hunk 5]
```diff
@@ -145,7 +145,7 @@ endfunction()
 if (PYBIND11_TEST)
   enable_testing()
   add_subdirectory(example)
-  add_custom_target(check COMMAND ${CMAKE_CTEST_COMMAND} -C $<CONFIGURATION> DEPENDS example)
+  add_custom_target(check COMMAND ${CMAKE_CTEST_COMMAND} -C $<CONFIGURATION> DEPENDS example example18)
```
Comment (wjakob) [inline CMakeLists.txt]: This won't be needed if example18 is a normal part of the extension library.

File: docs/advanced.rst [hunk 6]
```diff
@@ -1562,3 +1562,27 @@ work, it is important that all lines are indented consistently, i.e.:
...
+
+Calling Python from C++
+=======================
+
+Pybind11 also allows to call python code from C++.
+
+.. code-block:: cpp
+
+    // initialize the interpreter.
+	Py_Initialize() ;
```
Comment (wjakob) [inline docs/advanced.rst]: No Py_Initialize() please.

File: docs/advanced.rst [hunk 7]
```diff
@@ -1562,3 +1562,27 @@ work, it is important that all lines are indented consistently, i.e.:
 
 .. [#f4] http://www.sphinx-doc.org
 .. [#f5] http://github.com/pybind/python_example
+
+Calling Python from C++
+=======================
+
+Pybind11 also allows to call python code from C++.
```
Comment (wjakob) [inline docs/advanced.rst]: More complete documentation would be desirable, see e.g. http://www.boost.org/doc/libs/1_38_0/libs/python/doc/v2/exec.html

File: example/example18.cpp [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+
+    All rights reserved. Use of this source code is governed by a
+    BSD-style license that can be found in the LICENSE file.
+*/
+
+#include "example.h"
+#include <pybind11/exec.h>
+
+int main(int argc, char* argv[])
+{
```
Comment (wjakob) [inline example/example18.cpp]: Indentation issue

**Comment by wjakob** [comment]
I added a few more comments -- it will be ready to merge once those are addressed.

File: example/example18.cpp (hunk 4 above)
Comment (klemens-morgenstern) [inline example/example18.cpp]: Sry, I am bad at using cmake. When I add example18 to PYBIND11_EXAMPLES I'll get a library, but I need an executable. I don't know how that should work.

File: docs/advanced.rst (hunk 6 above)
Comment (klemens-morgenstern) [inline docs/advanced.rst]: As in removing it or as in hiding it?

File: example/example18.cpp (hunk 4 above)
Comment (wjakob) [inline example/example18.cpp]: That is how it is supposed to be. Just add a test() function that becomes part of the library, and which you call from python (and which in turn calls back to Python via eval()).

File: docs/advanced.rst (hunk 6 above)
Comment (wjakob) [inline docs/advanced.rst]: You won't need it if example18 is part of the extension module, as are example1-17.

**Comment by klemens-morgenstern** [comment]
Ok, I fixed the test execution, added a bit documentation, i.e. a complete example and a bit more info to the reference. Hope that addresses all.

**Comment by wjakob** [comment]
One more comment -- it seems a bit odd to write "Blatantly stolen" and then in the next line " Copyright (c) 2016 Klemens D. Morgenstern <klemens.morgenstern@gmx.net>" ;)

**Comment by klemens-morgenstern** [comment]
Well ok, that's true. Well yeah, since the licesnse is the same and I just matched it, copyright would remain with the original authors. I'll fix that.

**Comment by wjakob** [comment]
Great, thank you. Any ideas why the test suite (specifically, Linux 64 bit / GCC 4.8) fails on your new testcase?

File: include/pybind11/exec.h [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+    if (res == nullptr)
+        throw error_already_set();
+
+    return {res, false};
+}
+
+inline object exec_file(str filename, object global = object(), object local = object()) {
+    // Set suitable default values for global and local dicts.
+    if (!global) {
+    if (PyObject *g = PyEval_GetGlobals())
```
Comment (aldanor) [inline include/pybind11/exec.h]: Minor: indentation

File: include/pybind11/exec.h [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+
+    if (res == nullptr)
+        throw error_already_set();
+
+    return {res, false};
+
+}
+
+inline object exec (const std::string &string, object global = object(), object local = object()) {
+    return exec(str(string), global, local);
```
Comment (aldanor) [inline include/pybind11/exec.h]: Minor, but here you convert `string -> str -> string`, would it make sense to do it the other way around and avoid double cast?

**Comment by klemens-morgenstern** [comment]
@wjakob I have no clue, it does work with my 64-Bit VM and it did work as a C++ case.

File: include/pybind11/exec.h (hunk 10 above)
Comment (klemens-morgenstern) [inline include/pybind11/exec.h]: Yes, that would make sense. I wanted to have the default variant so be pybind::str, and then I quickly added the std::string version to allow string literals. I'll change that.

**Comment by wjakob** [comment]
FYI: I wasn't really happy with the Boost.Python design and decided to rewrite it from scratch (but I kept your testcases). So this feature is finally available.

**Pull request closed without merging**